plugins/ring/ring.ml

```diff
@@ -431,56 +431,56 @@ let op_smorph r add mul req m1 m2 =
-                        str"\"");
-                  op_morph)
-            | None ->
-                (Flags.if_verbose
-                    Feedback.msg_info
-                    (str"Using setoid \""++pr_econstr_env env sigma req ++str"\"" ++ spc() ++
-                        str"and morphisms \""++pr_econstr_env env sigma add_m ++
-                        str"\""++spc()++str"and \""++
-                        pr_econstr_env env sigma mul_m++str"\"");
-                 op_smorph r add mul req add_m_lem mul_m_lem) in
-          (setoid,op_morph)
+  | App (f, [| _ |]) when isRefX env sigma (rocq_eq ()) f ->
+    let sigma, setoid = plapp sigma rocq_eq_setoid [|r|] in
+    let sigma, op_morph =
+      match opp with
+        Some opp -> plapp sigma rocq_eq_morph [|r;add;mul;opp|]
+      | None -> plapp sigma rocq_eq_smorph [|r;add;mul|] in
+    let sigma, setoid = Typing.solve_evars env sigma setoid in
+    let sigma, op_morph = Typing.solve_evars env sigma op_morph in
+    (sigma,setoid,op_morph)
+  | _ ->
+    let sigma, setoid = setoid_of_relation env sigma r req in
+    let signature = [Some (r,Some req);Some (r,Some req)],Some(r,Some req) in
+    let sigma, add_m, add_m_lem =
+      try Rewrite.Internal.default_morphism env sigma signature add
+      with Not_found ->
+        CErrors.user_err (str "Ring addition " ++ pr_econstr_env env sigma add ++ str " should be declared as a morphism.") in
+    let sigma, mul_m, mul_m_lem =
+      try Rewrite.Internal.default_morphism env sigma signature mul
+      with Not_found ->
+        CErrors.user_err (str "Ring multiplication " ++ pr_econstr_env env sigma mul ++ str " should be declared as a morphism.") in
+    let sigma, op_morph =
+      match opp with
+      | Some opp ->
+        (let sigma, opp_m,opp_m_lem =
+           try Rewrite.Internal.default_morphism env sigma ([Some(r,Some req)],Some(r,Some req)) opp
+           with Not_found ->
+             CErrors.user_err (str "Ring opposite " ++ pr_econstr_env env sigma opp ++ str " should be declared as a morphism.") in
+         let op_morph =
+           op_morph r add mul opp req add_m_lem mul_m_lem opp_m_lem in
+         Flags.if_verbose
+           Feedback.msg_info
+           (str"Using setoid \""++ pr_econstr_env env sigma req++str"\""++spc()++
+            str"and morphisms \""++pr_econstr_env env sigma add_m ++
+            str"\","++spc()++ str"\""++pr_econstr_env env sigma mul_m++
+            str"\""++spc()++str"and \""++pr_econstr_env env sigma opp_m++
+            str"\"");
+         sigma, op_morph)
+      | None ->
+        (Flags.if_verbose
+           Feedback.msg_info
+           (str"Using setoid \""++pr_econstr_env env sigma req ++str"\"" ++ spc() ++
+            str"and morphisms \""++pr_econstr_env env sigma add_m ++
+            str"\""++spc()++str"and \""++
+            pr_econstr_env env sigma mul_m++str"\"");
+         sigma, op_smorph r add mul req add_m_lem mul_m_lem) in
+    (sigma,setoid,op_morph)
```

Comment:
This sigma-threading business seems like an unrelated fix, the code was incorrect already before. Could you split that in a separate commit?

Reply:
done